File: chat_hub/core/services/chat/message_service.go
package services

import (
	response_dtos "chat_hub/core/domain/dtos/response"
	entity "chat_hub/core/domain/entities"
	"chat_hub/infrastructure/repository"
	"database/sql"
	"fmt"
	"log"
	"strconv"
	"time"
)

type MessageService struct {
	db *sql.DB

	messageRepository *repository.MessageRepository
}

func NewMessageService(db *sql.DB) *MessageService {
	return &MessageService{
		db:                db,
		messageRepository: repository.NewMessageRepository(db),
	}
}

func (s *MessageService) CreateMessage(dialogue entity.UserDialogueEntity, userId, message, userMessageId string, senderType, messageType string) (string, error) {
	messageRequest, err := s.buildMessage(dialogue, userId, message, userMessageId, senderType, messageType)
	if err != nil {
		log.Println("Error building message: " + err.Error())
		return "", err
	}
	messageId, err := s.messageRepository.CreateMessage(messageRequest)
	if err != nil {
		log.Println("Error creating message: " + err.Error())
		return "", err
	}
	log.Println("Message created with ID: " + messageId)
	return messageId, nil
}

func (s *MessageService) buildMessage(dialogue entity.UserDialogueEntity, userId, message, userMessageId string, senderType, messageType string) (entity.MessageEntity, error) {
	userIdF, err := strconv.ParseInt(userId, 10, 64)
	if err != nil {
		return entity.MessageEntity{}, fmt.Errorf("invalid user ID: %v", err)
	}
	entity := entity.MessageEntity{
		UserId:        userIdF,
		DialogueId:    dialogue.ID,
		UserMessageId: userMessageId,
		MessageType:   messageType,
		Content:       message,
		Metadata:      "",
		SenderType:    senderType,
		CreatedAt:     time.Now(),
		UpdatedAt:     time.Now(),
	}
	return entity, nil
}

func (s *MessageService) GetMessageByDialogueId(dialogueId string, numberOfMessages int) (response_dtos.RecentHistory, error) {
	recentChatMessages, err := s.messageRepository.GetRecentChatMessagesByDialogueId(dialogueId, numberOfMessages)
	if err != nil {
		log.Println("Error getting recent chat messages: " + err.Error())
		return response_dtos.RecentHistory{}, err
	}
	if len(recentChatMessages) == 0 {
		log.Println("No messages found for dialogue ID: " + dialogueId)
		return response_dtos.RecentHistory{}, nil
	}
	var recentChatMessagesDTO []response_dtos.MessageResponseDTO
	for _, message := range recentChatMessages {
		recentChatMessagesDTO = append(recentChatMessagesDTO, response_dtos.MessageResponseDTO{
			Message:    "<" + message.SenderType + "> " + message.Content,
			Metadata:   message.Metadata,
		})
	}

	var recentChatHistoryDTO response_dtos.RecentHistory
	recentChatHistoryDTO.UserId = recentChatMessages[0].UserId
	recentChatHistoryDTO.DialogueId = recentChatMessages[0].DialogueId
	recentChatHistoryDTO.Messages = recentChatMessagesDTO

	return recentChatHistoryDTO, nil
}

func (s *MessageService) GetMessageByPagination(dialogueId string, size int, cursor string) ([]entity.MessageEntity, error) {
	messages, err := s.messageRepository.GetMessagesByDialogueIdWithCursorPagination(dialogueId, size, cursor)
	if err != nil {
		log.Println("Error getting messages by pagination: " + err.Error())
		return nil, err
	}
	if len(messages) == 0 {
		log.Println("No messages found for dialogue ID: " + dialogueId)
		return nil, nil
	}
	return messages, nil
}

File: ai_core/core/service/integration/message_service.py
import uuid
import time

from core.domain.entities.user_dialogue import UserDialogue
from core.domain.entities.message import Message
from core.domain.request.chat_hub_request import RecentHistoryRequest
from core.domain.response.message_response import MessageResponseDTO, RecentHistory
from core.service.integration.dialogue_service import dialogue_service
from infrastructure.repository.message_repository import message_repository


class MessageService:
    """Service layer for messages, interfacing with the MessageRepository."""

    async def create_message(self, dialogue: UserDialogue, user_id: int, message: str, user_message_id: str, sender_type: str, message_type: str) -> str:
        message_request = self._build_message(
            dialogue, user_id, message, user_message_id, sender_type, message_type)
        message_id = await message_repository.create_message(message_request)
        print(f"Created message with ID: {message_id}")
        return message_id

    def _build_message(self, dialogue: UserDialogue, user_id: int, message: str, user_message_id: str, sender_type: str, message_type: str) -> Message:
        new_message = Message(
            id=uuid.uuid4(),
            user_id=user_id,
            dialogue_id=dialogue.id,
            user_message_id=str(user_message_id),
            message_type=message_type,
            content=message,
            sender_type=sender_type,
            metadata="",
        )
        return new_message

    async def get_message_by_dialogue_id(self, dialogue_id: str, number_of_messages: int) -> Message:
        print("Dialogue ID in get_message_by_dialogue_id:", dialogue_id)
        recent_chat_messages = await message_repository.get_recent_chat_messages_by_dialogue_id(dialogue_id, number_of_messages)
        if len(recent_chat_messages) == 0:
            print("No messages found for dialogue_id:", dialogue_id)
            return None
        recent_chat_messages_response = []
        for message in recent_chat_messages:
            recent_chat_messages_response.append(MessageResponseDTO(
                message="<" + message.sender_type + "> " + message.content,
                metadata=message.metadata
            ))

        recent_chat_his: RecentHistory = RecentHistory(
            user_id=recent_chat_messages[0].user_id,
            dialogue_id=dialogue_id,
            messages=recent_chat_messages_response
        )
        return recent_chat_his

    async def get_messages_by_dialogue_id_with_cursor_pagination(
        self, dialogue_id: str, size: int, cursor: str | None
    ) -> tuple[list[Message], bool]:
        messages, has_more = await message_repository.get_messages_by_dialogue_id_with_cursor_pagination(
            dialogue_id, size, cursor
        )
        if len(messages) == 0:
            print("No messages found for dialogue_id:", dialogue_id)
            return [], False
        # reverse messages to chronological order
        messages.reverse()
        return messages, has_more

    async def get_recent_history(self, request: RecentHistoryRequest) -> str:
        try:
            dialogue = await dialogue_service.get_dialogue_by_id(request.user_id, request.dialogue_id)
            if dialogue is None:
                print("Dialogue not found for dialogue_id:", request.dialogue_id)
                return ''
            result = await self.get_message_by_dialogue_id(dialogue.id, request.number_of_messages)
            if not result:
                print("No recent history found.")
                return ''
            if not self._validate_recent_history_response(request, result):
                print("Recent history response validation failed.")
                return ''

            return self._format_history(result.get('messages'))
        except Exception as e:
            print(f"Error in ChatHubServiceClient.get_recent_history: {e}")
            return ''

    def _validate_recent_history_response(self, request: RecentHistoryRequest, response: dict) -> bool:
        if int(request.user_id) != int(response.get('userId')):
            print(f"User ID mismatch: {request.user_id} != {response.get('userId')}")
            return False
        if request.dialogue_id != response.get('dialogueId'):
            print(f"Dialogue ID mismatch: {request.dialogue_id} != {response.get('dialogueId')}")
            return False
        return True

    def _format_history(self, history_list):
        lines = []
        for msg in history_list:
            if msg["message"].startswith("<user>"):
                lines.append("User: " + msg["message"].replace("<user>", "").strip())
            elif msg["message"].startswith("<bot>"):
                lines.append("GAIA: " + msg["message"].replace("<bot>", "").strip())
        return "\n".join(lines)

message_service = MessageService()
